[ECO-4122] v2.0.0 changelog

VeskeR opened the pull request:
Changelog is based on changes between https://github.com/ably/ably-js/compare/e71ed5febf98e7e73683088c7f5c98b2f5cb721b...c02ec568b51be10f0e4c70805f36a01514b8fcac (latest `main` and `integration/v2` diff as of March 21).

PRs that need to be added to the changelog when finished/reviewed and merged:
- [x] https://github.com/ably/ably-js/pull/1613 - breaking API change
- [x] https://github.com/ably/ably-js/pull/1633 - platform compatibility PR
- [x] https://github.com/ably/ably-js/pull/1640 - API change
- [x] https://github.com/ably/ably-js/pull/1645 - breaking API change, removes `xhr_streaming` Transport
- [x] https://github.com/ably/ably-js/issues/1621 - breaking API change
- [x] https://github.com/ably/ably-js/issues/1637 - breaking API change
- [x] https://github.com/ably/ably-js/pull/1663 - breaking API change
- [x] https://github.com/ably/ably-js/pull/1658 - new API feature
- [x] https://github.com/ably/ably-js/pull/1654 - `ChannelProvider` documentation
- [x] https://github.com/ably/ably-js/pull/1675 - falls under refactoring

Please note that this is not an exhaustive list of tasks that need to be completed before the v2 release. This list contains only the user-facing PRs/tasks that must be added to the changelog.

Pull requests not included in this changelog (I considered them as Miscellaneous / Extras, non-user facing):
- Use useLayoutEffect when calling .setOptions() in ChannelProvider [\#1706](https://github.com/ably/ably-js/pull/1706)
- Remove some lingering any from stats() param type [\#1705](https://github.com/ably/ably-js/pull/1705)
- Improve errors for incorrect constructor args [\#1702](https://github.com/ably/ably-js/pull/1702)
- Fix stray second arg to modular constructor [\#1701](https://github.com/ably/ably-js/pull/1701)
- Add migration guide for react hooks for ably-js v2 [\#1698](https://github.com/ably/ably-js/pull/1698)
- Change PresenceEnterResult to PresenceResult in react hooks [\#1697](https://github.com/ably/ably-js/pull/1697)
- Fix ably/modular can't be imported in node environment [\#1696](https://github.com/ably/ably-js/pull/1696)
- Fix lint warnings [\#1694](https://github.com/ably/ably-js/pull/1694)
- build: use esbuild for nodejs bundle [\#1691](https://github.com/ably/ably-js/pull/1691)
- Add documentation for `skip` parameter for hooks to React docs [\#1675](https://github.com/ably/ably-js/pull/1675)
- Add migration guide for v2 [\#1670](https://github.com/ably/ably-js/pull/1670)
- Document fallback transport mechanism in the context of network conditions for v2 [\#1664](https://github.com/ably/ably-js/pull/1664)
- Restore ClientOptions.plugins mechanism and use it for modular variant [\#1663](https://github.com/ably/ably-js/pull/1663)
- ci: don't check formatting in react workflow [\#1653](https://github.com/ably/ably-js/pull/1653)
- test: test ux improvements [\#1652](https://github.com/ably/ably-js/pull/1652)
- Move React.createContext outside of components in react-hooks [\#1643](https://github.com/ably/ably-js/pull/1643)
- Improve logs for react package test in test:package [\#1639](https://github.com/ably/ably-js/pull/1639)
- Document our continued support for v1 [\#1634](https://github.com/ably/ably-js/pull/1634)
- Refactor nodejs http util class [\#1627](https://github.com/ably/ably-js/pull/1627)
- Fix missing ably/react export from ably package on v2 branch [\#1625](https://github.com/ably/ably-js/pull/1625)
- Fix ably import in react hooks [\#1624](https://github.com/ably/ably-js/pull/1624)
- Convert HTTP code to use promises [\#1603](https://github.com/ably/ably-js/pull/1603)
- Add better error handling for incorrect PLAYWRIGHT_BROWSER in runPlaywrightTests script [\#1595](https://github.com/ably/ably-js/pull/1595)
- Document that queryTime auth option requires Rest module [\#1593](https://github.com/ably/ably-js/pull/1593)
- Make modulereport print gzipped size too [\#1586](https://github.com/ably/ably-js/pull/1586)
- Add try/catch to _multiple_send to allow fast fail in case of errors [\#1585](https://github.com/ably/ably-js/pull/1585)
- Fix mocha web server process keeps running after failed playwright tests [\#1584](https://github.com/ably/ably-js/pull/1584)
- Fix multiple_send_binary_2_200 is failing intermittently in browser on v2 branch [\#1583](https://github.com/ably/ably-js/pull/1583)
- Improve IPlatformConfig [\#1563](https://github.com/ably/ably-js/pull/1563)
- build: use Ably as UMD lib name [\#1558](https://github.com/ably/ably-js/pull/1558)
- Remove callbacks part 1 [\#1541](https://github.com/ably/ably-js/pull/1541)
- Strip logging from modular variant of the SDK [\#1536](https://github.com/ably/ably-js/pull/1536)
- Correct tsc module settings for test:package [\#1521](https://github.com/ably/ably-js/pull/1521)
- Add stats to list of functionality gated off by Rest module [\#1517](https://github.com/ably/ably-js/pull/1517)
- Remove the existing plugins mechanism [\#1512](https://github.com/ably/ably-js/pull/1512)
- Add some regression testing for bundle size and included files [\#1507](https://github.com/ably/ably-js/pull/1507)
- Exclude PresenceMessage from BaseRealtime bundle [\#1505](https://github.com/ably/ably-js/pull/1505)
- Remove static stuff from *Message classes [\#1503](https://github.com/ably/ably-js/pull/1503)
- Add documentation for modular variant of library [\#1502](https://github.com/ably/ably-js/pull/1502)
- Add typings for the tree-shakable version of the library [\#1498](https://github.com/ably/ably-js/pull/1498)
- Remove remaining REST code from BaseRealtime [\#1496](https://github.com/ably/ably-js/pull/1496)
- Don’t populate DeviceDetails.toRequestBody by assignment [\#1494](https://github.com/ably/ably-js/pull/1494)
- Do e013cc0’s typescript-eslint version bumps in package.json [\#1490](https://github.com/ably/ably-js/pull/1490)
- Add a basic test of NPM package [\#1476](https://github.com/ably/ably-js/pull/1476)
- Make MessageFilter subscription filtering tree-shakable[\#1473](https://github.com/ably/ably-js/pull/1473)
- Make HTTP request implementations tree-shakable [\#1438](https://github.com/ably/ably-js/pull/1438)
- Make realtime transports tree-shakable [\#1432](https://github.com/ably/ably-js/pull/1432)
- Remove PresenceMessage-related static things in tree-shakable library [\#1429](https://github.com/ably/ably-js/pull/1429)
- Make realtime presence functionality tree-shakable [\#1428](https://github.com/ably/ably-js/pull/1428)
- Create tree-shakable MsgPack module [\#1425](https://github.com/ably/ably-js/pull/1425)
- Create tree-shakable Crypto module [\#1419](https://github.com/ably/ably-js/pull/1419)
- Add tree-shakable Base* and Rest exports [\#1417](https://github.com/ably/ably-js/pull/1417)
- refactor: convert web comet transports to typescript [\#1379](https://github.com/ably/ably-js/pull/1379)
- build: use esbuild for web bundles [\#1369](https://github.com/ably/ably-js/pull/1369)
- Update the tests to use the Promise-based API [\#1349](https://github.com/ably/ably-js/pull/1349)
- Make ErrorInfo.fromValues’s signature correspond to its expectations [\#1322](https://github.com/ably/ably-js/pull/1322)
- Fix TypeScript 4.9.5 upgrade mistakes [\#1267](https://github.com/ably/ably-js/pull/1267)
- Upgrade TypeScript to 4.9.5 [\#1258](https://github.com/ably/ably-js/pull/1258)
- Update docstrings for Rest.request version param [\#1255](https://github.com/ably/ably-js/pull/1255)
- Convert platform crypto.js files to TypeScript [\#1246](https://github.com/ably/ably-js/pull/1246)
- Upgrade webpack to version 5 [\#1201](https://github.com/ably/ably-js/pull/1201)

Resolves #1604

owenpearson commented on CHANGELOG.md:
```diff
@@ -2,6 +2,85 @@
+- Remove `ably-commonjs*.js` files [\#1229](https://github.com/ably/ably-js/pull/1229)
+- Remove remaining deprecated APIs [\#1227](https://github.com/ably/ably-js/pull/1227)
+- Remove deprecated `fromEncoded*` type declarations [\#1222](https://github.com/ably/ably-js/pull/1222)
+- Conform to spec for logging configuration [\#1216](https://github.com/ably/ably-js/pull/1216)
```
I think we should avoid mentioning the spec in changelog entries whenever possible, feel free to editorialise some of the PR titles

VeskeR replied:
I assume this is because "the spec" term is not something that users are familiar with, and there is no point to expose them to it, right?
So a more user-friendly explanation on **what** actually was done would be much more useful.

VeskeR replied:
see https://github.com/ably/ably-js/pull/1656/commits/dd8db9409cb3cde90bfc321ddd33ea71bd3009c2

owenpearson commented on CHANGELOG.md:
```diff
@@ -2,6 +2,85 @@
+- `ChannelProvider` implementation [\#1620](https://github.com/ably/ably-js/pull/1620)
+- Remove duplicate type names and change public `ChannelModes` type [\#1601](https://github.com/ably/ably-js/pull/1601)
+- Remove `any` from `stats()` param type [\#1561](https://github.com/ably/ably-js/pull/1561)
+- Make modular variant of the library available to Web Workers [\#1550](https://github.com/ably/ably-js/pull/1550)
```
This title doesn't really describe a breaking change - the real breaking change is that we've removed the web-worker specific build

VeskeR replied:
Agree, will change

VeskeR replied:
see https://github.com/ably/ably-js/pull/1656/commits/dd8db9409cb3cde90bfc321ddd33ea71bd3009c2

owenpearson reviewed (commented):
Looks good on the whole, I agree with the categorisation of PRs into non-user facing, breaking, etc.

I do think, if this is gonna be what ends up as the description in the github release we should be a bit more 'product marketing' about how we present the changes at first, and maybe even start off with a short paragraph rather than going straight into bullet points. It doesn't look too exciting if the first thing someone sees is that we've dropped support for platforms, whereas if we start by saying stuff like "We've added a new way to remove bloat from your ably-js client and we're adding first class support to promises" that sounds better IMO.

A couple of things that stood out to me below:

VeskeR commented:
> I do think, if this is gonna be what ends up as the description in the github release we should be a bit more 'product marketing' about how we present the changes at first, and maybe even start off with a short paragraph rather than going straight into bullet points. It doesn't look too exciting if the first thing someone sees is that we've dropped support for platforms, whereas if we start by saying stuff like "We've added a new way to remove bloat from your ably-js client and we're adding first class support to promises" that sounds better IMO.

Updated intro.

I've put platform support deprecations first thinking that it might be useful to immediately see what would break if you switch to this new version.
Upon a second look, however, I do agree that it looks underwhelming, so it has been changed now.

lawrence-forooghian commented:
I think this needs to be reviewed as a single thing alongside the migration guide, the writing of which is assigned to me. I’m going to branch from this and write the migration guide, then we can review the two together.

lawrence-forooghian commented:
Started working on migration guide in https://github.com/ably/ably-js/pull/1670.

ttypic commented on CHANGELOG.md:
```diff
@@ -2,6 +2,105 @@
+
+### Modular variant of the library
+
+An ESM variant of the library is now available for browsers (but not for Node.js) at `ably/modules`. This modular variant of the library supports tree shaking, allowing for a reduction in the Ably bundle size within your application. It can also be used by Web Workers.
```
Now it's `ably/modular`

VeskeR replied:
changed in https://github.com/ably/ably-js/pull/1656/commits/a57271c43e82d4be53e081dcaa0e33d054026e49

ttypic commented on CHANGELOG.md:
```diff
@@ -2,6 +2,105 @@
+
+### React Hooks
+
+React hooks, exported at `ably/react`, now rely on the `ChannelProvider` component to define the channels you wish to use and the options for them.
```
Comment to add `usePresence` changes, when merged

VeskeR replied:
done in https://github.com/ably/ably-js/pull/1656/commits/a57271c43e82d4be53e081dcaa0e33d054026e49

VeskeR replied:
Added in https://github.com/ably/ably-js/pull/1656/commits/23dd825b4ae6d2b153417c7cea0f319bc2ce2d84

ttypic reviewed (commented):
Looks great!

VeskeR commented:
Added changelog for new changes since last update in this PR: https://github.com/ably/ably-js/compare/2462e9bfa412284a3c3975ad83a8ea4da9d9e7f4...85b54ea50cf040f061e1c6ff41445911741a522e :

- https://github.com/ably/ably-js/pull/1676
- https://github.com/ably/ably-js/pull/1658
- https://github.com/ably/ably-js/pull/1641
- https://github.com/ably/ably-js/pull/1640
- https://github.com/ably/ably-js/pull/1633
- https://github.com/ably/ably-js/pull/1613

Non-user facing (were not added to changelog doc):

- https://github.com/ably/ably-js/pull/1664
- https://github.com/ably/ably-js/pull/1663
- https://github.com/ably/ably-js/pull/1654
- https://github.com/ably/ably-js/pull/1643
- https://github.com/ably/ably-js/pull/1634

Also moved [[SDK-3938] Remove the existing plugins mechanism](https://github.com/ably/ably-js/pull/1512) to non-user facing, since it got reverted by [[ECO-4658] Restore ClientOptions.plugins mechanism and use it for modular variant](https://github.com/ably/ably-js/pull/1663)

VeskeR commented:
Updated PR to be based on latest changes in `integration/v2`: 85b54ea50cf040f061e1c6ff41445911741a522e...a29e83ff980f46206339d6001d6b77e74ee02009.
Nothing to add to the changelog, only merged PRs from main

VeskeR commented:
Added changelog for new changes since last update in this PR: https://github.com/ably/ably-js/compare/a29e83ff980f46206339d6001d6b77e74ee02009...779516bf11aa6f312b2f01a0b0ed5154d718e9a1:
- https://github.com/ably/ably-js/pull/1674

Not public API related (were not added to changelog doc):
- https://github.com/ably/ably-js/pull/1697
- https://github.com/ably/ably-js/pull/1696
- https://github.com/ably/ably-js/pull/1675
- https://github.com/ably/ably-js/pull/1652

VeskeR commented on CHANGELOG.md:
```diff
@@ -2,6 +2,110 @@
+
+The 2.0.0 release introduces a number of new features and QoL improvements, including a new way to remove bloat and reduce the bundle size of your ably-js client, first-class support for Promises, a more idiomatic approach to using ably-js' React Hooks, enhancements to TypeScript typings, and more.
+
+Below is a list of all new features, APIs, deprecations, and breaking changes.
```
I don't think that overlap by itself is a bad thing.
Also, as it currently stands, the migration guide is very "raw" and goes immediately into the code details on how to migrate to v2, which I believe it should do.
The changelog lists all the things that were added, changed, and improved, while the migration guide explains step by step how to migrate. So, I don't think there is a real overlap here.

> highlight the most important breaking changes (e.g. platform compatibility) and direct users to the migration guide for the full details?

Changelog already has a list of most important features listed in separate headings, and the link to the migration guide is on the line below. Only after that we list all of the PRs one by one.
Did you have something else in mind on how to do this?
The changelog already lists the most important features under separate headings below (before listing PRs), and the link to the migration guide is provided on the line below.
Did you have something else in mind on how to do this?

lawrence-forooghian replied:
> The changelog lists all the things that were added, changed, and improved

Maybe it would be good to organise the prose part of the changelog into sections that align with these categories?

> I don't think that overlap by itself is a bad thing.

Agreed, I think I was not very careful in my comment before. The aim is to have a set of documentation that's useful to users, and that might involve repeating ourselves.

VeskeR replied:
> Maybe it would be good to organise the prose part of the changelog into sections that align with these categories?

If by `prose part` you mean those high-level topics like "Bundle Size Reduction" and "React Hooks", I feel like it doesn't really make sense for them. It's better to list them as is, by part of the application that was impacted and explain what's changed for each of them.

For the PRs though, I reordered them so they form those categories.

lawrence-forooghian replied:
My thinking was that it would be nice for a user who looks at the changelog to know which stuff they _need_ to pay attention to (because else their code will break) and which stuff they could read later. But up to you.

VeskeR replied:
Reordered PRs in https://github.com/ably/ably-js/pull/1656/commits/f0995b4792a7dbcb73dd4e19a22412f8ee90b073

VeskeR commented on CHANGELOG.md:
```diff
@@ -2,6 +2,110 @@
+
+___
+
+### Breaking Changes
```
In general, I don't think we should "hide" information if it can be easily displayed. We have split the changelog into sections so that users who are interested in the high-level view can read only the first part, and if they want more specific information, they can read about merged PRs.

> refactorings: I'm not sure why users of the library would be interested in these internal changes

Even though some of the refactorings made are breaking changes, I would agree that it might not be necessary to list them separately again in the "Code Refactoring" section - one mention in the "Breaking Changes" section probably is enough.

> features and breaking changes

I think it's beneficial to list them all in a compressed format, one by one. The high-level overview above does not go into the specifics of how or why some things work, and those PRs provide quick access to that information. The migration guide, on the other hand, is very low-level, and there is no way to have a really quick look at it and get a feel for the required changes to migrate to v2.

lawrence-forooghian replied:
> In general, I don't think we should "hide" information if it can be easily displayed

I mean, if you think it's beneficial to include these lists of PRs in the changelog, I guess we can, but I would suggest that they be de-emphasised, perhaps by putting them inside `<details>` tags and making it clear that what follows is a list of PRs included in the changelog.

As it currently stands, the list of PRs looks like something that the user is _expected_ to read through in order to understand what's changed in v2 (especially since this changelog entry is likely to be a user’s first point of contact with v2). I can't speak for all users, but I certainly would not enjoy being expected to read through a bunch of PRs upon upgrading to a new version of a library that I use.

Bear in mind the intended audience of a pull request; it's not aimed at users, but rather at other developers in the team, who have knowledge about the context in which the PR is being created which a regular user wouldn’t have (or be expected to have). Furthermore, a single PR doesn't necessarily represent a single user-facing change; sometimes a change is split up into several PRs (possibly being performed iteratively, not necessarily getting it right the first time; software development is messy), and sometimes a PR contains a bunch of unrelated changes, some of which are relevant to users, some of which aren't (e.g. refactors). And, then, once a PR is created, it goes through a bunch of feedback and changes, meaning that the final state of a PR might not even fully reflect the changes described in its title or description, so to fully understand the impact of a PR a user would need to then look at all this stuff too.

> I think it's beneficial to list them all in a compressed format, one by one. The high-level overview above does not go into the specifics of how or why some things work, and those PRs provide quick access to that information. 

The fact the list of PRs appears concise doesn't necessarily imply that it's understandable to users.

> The migration guide, on the other hand, is very low-level

I’m not sure what you mean by "low-level" here. It's written in prose and explicitly aimed at users of the library. How is this more "low-level" than PRs which, ultimately, show the _changes to the library’s implementation_?

lawrence-forooghian replied:
So, to summarise — I think that we can include a list of PRs as "bonus" information for particularly interested users to view. But I think that any information that we consider important to communicate in this changelog should be communicated as prose that's explicitly targeted at our users.

lawrence-forooghian replied:
> some of the refactorings made are breaking changes

What's your definition of a "refactoring" here? My understanding of a refactoring is that it's a change that has no user-facing impact, i.e. it could never be a breaking change.

VeskeR replied:
> by putting them inside `details` tags and making it clear that what follows is a list of PRs included in the release.

> As it currently stands, the list of PRs looks like something that the user is expected to read through in order to understand what's changed in v2

> so to fully understand the impact of a PR a user would need to then look at all this stuff too.

Agree, I never intended for users to go through each PR and figure out what's changed - this is why we have a thorough migration guide that just tells you what to do. I've put PRs in the details tag as per your suggestion.

> But I think that any information that we consider important to communicate in this changelog should be communicated as prose that's explicitly targeted at our users.

Agree, this is how changelog looks like after hiding PRs in the details. Will also change wording in couple of sections to be less code-specific (like https://github.com/ably/ably-js/pull/1656#discussion_r1530506245 below)

VeskeR replied:
> What's your definition of a "refactoring" here? My understanding of a refactoring is that it's a change that has no user-facing impact, i.e. it could never be a breaking change.

Probably not the best choice of a word here. Refactorings are changes that restructure the code in some way, but do not alter its functionality. However even a change like this can be a breaking change: changing a name of a prop in some object which is part of the public API - looks like refactoring, but is still a breaking change. On this note, you could argue that changing public API **is** changing functionality in the library context - maybe, but it does not feel like a "feature" or a "bug fix" either. So I just put them under refactorings section.

Now, based on what was discussed above, I do feel now that there is no need to be so verbose when listing those PRs.
Currently, all user-facing PRs are split into three sections: "features", "bug fixes" and "refactorings", and then if any PR is a breaking change, it is then **also** listed in the breaking changes section.

I've changed that to list each PR only once, either in "breaking changes", "features" or "bug fixes" and removed "refactorings" section altogether.

lawrence-forooghian replied:
Yeah — changing a property name in the public API is a breaking change, not a refactoring. The new headings that you've suggested seem sensible to me.

lawrence-forooghian commented on CHANGELOG.md:
```diff
@@ -2,6 +2,122 @@
+
+### First-class support for Promises
+
+The exports `ably/callbacks` and `ably/promises` have been removed. The callbacks API has been entirely removed, and now only promises are supported. Calls like `new Ably.Realtime.Promise` should be replaced with `new Ably.Realtime`.
```
Bearing in mind your approach described [above](https://github.com/ably/ably-js/pull/1656/files#r1529432940):

> The changelog lists all the things that were added, changed, and improved, while the migration guide explains step by step how to migrate.

…it probably would be best not to include this instruction, since it's not a very useful one in isolation and requires the migration guide to properly understand how a user needs to change their code. What do you think?

VeskeR replied:
Aggree, reworded to describe the change more generally and don't include specific instructions

VeskeR commented:
Added changelog for new changes since last update in this PR: https://github.com/ably/ably-js/compare/779516bf11aa6f312b2f01a0b0ed5154d718e9a1...2e124597417204ff6b548502b756adf490f19c17:
- https://github.com/ably/ably-js/pull/1645

Not public API related (were not added to changelog doc):
- https://github.com/ably/ably-js/pull/1705
- https://github.com/ably/ably-js/pull/1702
- https://github.com/ably/ably-js/pull/1701
- https://github.com/ably/ably-js/pull/1691

ttypic reviewed (approved):
Looks good

lawrence-forooghian commented on CHANGELOG.md:
```diff
@@ -2,6 +2,92 @@
+
+### Bundle Size Reduction
+
+The default bundle size for the web has been reduced by 30% compared to v1 - from 231.35 KiB to 162.15 KiB. When calculated with gzip compression, the reduction is 28.5%, from 81.68 KiB to 58.48 KiB.
```
Just to check — are these numbers up to date for latest v1 and v2 (especially after the removal of `xhr_streaming` in v2)?

VeskeR replied:
Good point!
Updated in https://github.com/ably/ably-js/pull/1656/commits/9b908a0942a29477ea75f702f8931f99c22f2593.
v1 actually increased a bit in size (those deprecation messages come at a cost), but v2 decreased thanks to removal of `xhr_streaming` and connection upgrade code

lawrence-forooghian replied:
Awesome! Feel free to merge 🙂

VeskeR commented:
Updated for latest changes https://github.com/ably/ably-js/compare/2e124597417204ff6b548502b756adf490f19c17...c02ec568b51be10f0e4c70805f36a01514b8fcac:

no user-facing changes
